contracts/zaps/ZapSteth.sol

```diff
@@ -0,0 +1,321 @@
+        uint256 balance2;
+        //test if we should buy instead of mint
+        uint256 out = StableSwapSTETH.get_dy(0, 1, halfBal);
+        if (out < halfBal) {
+            balance2 = stETH.submit{ value: halfBal }(owner);
+        }
+
+        uint256 balanceMid = address(this).balance;
+
+        //add liquidity with no minimum mint
+        StableSwapSTETH.add_liquidity{ value: balanceMid }(
+            [balanceMid, balance2],
+            0
+        );
+
+        uint256 outAmount = want.balanceOf(address(this));
+
+        want.transfer(address(_position), outAmount);
+
+        (uint256 ptMinted, uint256 ytMinted) = tranche.prefundedDeposit(
+            msg.sender
+        );
+
+        require(ytMinted >= outAmount, "Not enough YT minted");
+        require(ptMinted >= _ptExpected, "Not enough PT minted");
+        return (ptMinted, ytMinted);
+    }
+
+    /// @notice This function takes stETH and converts it into yield and principal tokens
+    /// of a specified tranche.
+    /// @param _amount Amount of stETH to convert.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _ptExpected The minimum amount of principal tokens to mint.
+    /// @return returns the minted amounts of principal and yield tokens (PT and YT)
+    function zapStEthIn(
+        uint256 _amount,
+        uint256 _expiration,
+        address _position,
+        uint256 _ptExpected
+    ) external reentrancyGuard notFrozen returns (uint256, uint256) {
+        require(_amount != 0, "0 stETH");
+        ITranche tranche = _deriveTranche(address(_position), _expiration);
+        stETH.transferFrom(msg.sender, address(this), _amount);
+
+        uint256 balanceBegin = stETH.balanceOf(address(this));
+
+        // add liquidity with no minimum mint
+        StableSwapSTETH.add_liquidity([0, balanceBegin], 0);
+
+        uint256 outAmount = want.balanceOf(address(this));
+
+        want.transfer(address(_position), outAmount);
+
+        (uint256 ptMinted, uint256 ytMinted) = tranche.prefundedDeposit(
+            msg.sender
+        );
+
+        require(ytMinted >= outAmount, "Not enough YT minted");
+        require(ptMinted >= _ptExpected, "Not enough PT minted");
+        return (ptMinted, ytMinted);
+    }
+
+    /// @notice This function takes principal or yield tokens and redeems them for ETH.
+    /// @param _expiration Expiration of the target tranche.
+    /// @param _position The wrapped position contract address.
+    /// @param _amountPt Amount of principal tokens to redeem.
+    /// @param _amountYt Amount of yield tokens to redeem.
+    /// @param _outputExpected Minimum amount of ETH or STETH to return
+    function zapOutEth(
+        uint256 _expiration,
+        address _position,
+        uint256 _amountPt,
+        uint256 _amountYt,
+        uint256 _outputExpected
+    ) external reentrancyGuard notFrozen {
```

Comment:
We should add the permit hook from the user proxy to the two output functions. This will let us compress their usage to one txn in the front end.